roofit/roofitcore/test/testRooFuncWrapper.cxx

```diff
@@ -476,7 +480,105 @@ FactoryTestParams param5{"SimPdf", getSimPdfModel,
+
+   RooGaussian gauss("gauss", "gauss", x, mean, sigma);
+   RooRealVar frac("frac", "faction of histpdf", 0.5, 0, 1);
+   RooAddPdf model("model", "model", {histpdf, gauss}, frac);
+
+   ws.import(model);
+   ws.defineSet("observables", {x});
+}
+} // namespace
+
+/// Test based on rf706 tutorial
+FactoryTestParams param6{"HistPdf", getDataHistModel,
+                         [](RooAbsPdf &pdf, RooAbsData &data, RooWorkspace &) {
+                            return std::unique_ptr<RooAbsReal>{pdf.createNLL(data, RooFit::BatchMode("cpu"))};
+                         },
+                         1e-4,
+                         /*randomizeParameters=*/true};
+
+namespace {
+void getDataHistFuncModel(RooWorkspace &ws)
+{
+   using namespace RooFit;
+
+   int nEvents = 1000;
+   int numBins = 5;
+
+   ws.factory("x[0, 0, " + std::to_string(3 * numBins) + "]");
+
+   auto &x = *ws.var("x");
+   x.setBins(numBins);
+
+   // The parameters represent the expected events in each bin
+   RooArgList params = ParamHistFunc::createParamSet(ws, "gamma", x, 0, 1000);
+   for (auto *param : static_range_cast<RooRealVar *>(params)) {
+      param->setVal(nEvents / numBins);
+   }
+   ParamHistFunc phf{"phf", "phf", x, params};
+   RooRealSumPdf pdf{"pdf", "pdf", phf, RooArgList{1.0}};
+
+   // Enable the binned likelihood optimization to avoid integrals
+   // (like in HistFactory).
+   pdf.setAttribute("BinnedLikelihood");
+
+   ws.import(pdf);
+
+   // We have to wrap the pdf in a RooSimultaneous, otherwise the
+   // BinnedLikelihood optimization doesn't work.
+   ws.factory("SIMUL::model( cat[A=0], A=pdf)");
+   auto &model = *ws.pdf("model");
+   auto &cat = *ws.cat("cat");
+
+   std::unique_ptr<RooDataHist> data{model.generateBinned({x, cat}, nEvents)};
+   data->SetName("data");
+   ws.import(*data);
+
+   ws.defineSet("observables", *model.getObservables(*data));
+}
+} // namespace
+
+FactoryTestParams param7{"HistFuncPdf", getDataHistFuncModel,
+                         [](RooAbsPdf &pdf, RooAbsData &data, RooWorkspace &) {
+                            return std::unique_ptr<RooAbsReal>{pdf.createNLL(data, RooFit::BatchMode("cpu"))};
+                         },
+                         1e-4,
+                         /*randomizeParameters=*/true};
+
+FactoryTestParams param8{"Poisson",
+                         [](RooWorkspace &ws) {
+                            ws.factory("sum::mu_shifted(mu[0, -10, 10], shift[1.0, -10, 10])");
+                            ws.factory("Poisson::model(x[5, -10, 10], mu_shifted)");
+
+                            ws.defineSet("observables", "x");
+                         },
+                         [](RooAbsPdf &pdf, RooAbsData &data, RooWorkspace &) {
+                            return std::unique_ptr<RooAbsReal>{pdf.createNLL(data, RooFit::BatchMode("cpu"))};
+                         },
+                         1e-4,
+                         /*randomizeParameters=*/false};
```

Comment:
```suggestion
FactoryTestParams param8{"Poisson",
                         [](RooWorkspace &ws) {
                            ws.factory("sum::mu_shifted(mu[0, -10, 10], shift[1.0, -10, 10])");
                            ws.factory("Poisson::model(x[5, -10, 10], mu_shifted)");

                            ws.defineSet("observables", "x");
                         },
                         [](RooAbsPdf &pdf, RooAbsData &data, RooWorkspace &) {
                            return std::unique_ptr<RooAbsReal>{pdf.createNLL(data, RooFit::BatchMode("cpu"))};
                         },
                         1e-4,
                         /*randomizeParameters=*/false};


// A RooPoisson where x is not rounded, like it is used in HistFactory
FactoryTestParams param9{"PoissonNoRounding",
                         [](RooWorkspace &ws) {
                            ws.factory("sum::mu_shifted(mu[0, -10, 10], shift[1.0, -10, 10])");
                            ws.factory("Poisson::model(x[5, -10, 10], mu_shifted)");
                            auto poisson = static_cast<RooPoisson*>(ws.pdf("model"));
                            poisson->setNoRounding(true);

                            ws.defineSet("observables", "x");
                         },
                         [](RooAbsPdf &pdf, RooAbsData &data, RooWorkspace &) {
                            return std::unique_ptr<RooAbsReal>{pdf.createNLL(data, RooFit::BatchMode("cpu"))};
                         },
                         1e-4,
                         /*randomizeParameters=*/false};
```
It would be good to test both the case of rounding and no rounding (the no rounding case is maybe even more important because it is used in HistFactory that we want to support.